src/audio/asrc/asrc_common.c

```diff
@@ -0,0 +1,799 @@
+					  (int16_t **)cd->ibuf, &in_frames,
+					  (int16_t **)cd->obuf, &out_frames,
+					  &idx, 0);
+	else
+		ret = asrc_process_pull16(dev, cd->asrc_obj,
+					  (int16_t **)cd->ibuf, &in_frames,
+					  (int16_t **)cd->obuf, &out_frames,
+					  in_frames, &idx);
+
+	if (ret)
+		comp_err(dev, "src_copy_s16(), error %d", ret);
+
+	buf = (int16_t *)cd->obuf[0];
+	n = out_frames * audio_stream_get_channels(sink);
+	while (n > 0) {
+		n_wrap_snk = (int16_t *)audio_stream_get_end_addr(sink) - snk;
+		n_copy = (n < n_wrap_snk) ? n : n_wrap_snk;
+		s_copy = n_copy * sizeof(int16_t);
+		ret = memcpy_s(snk, s_copy, buf, s_copy);
+		assert(!ret);
+
+		/* Update and check both source and destination for wrap */
+		n -= n_copy;
+		snk += n_copy;
+		buf += n_copy;
+		src_inc_wrap_s16(&snk, audio_stream_get_end_addr(sink),
+				 audio_stream_get_size(sink));
+	}
+
+	*n_read = in_frames;
+	*n_written = out_frames;
+}
+
+static int asrc_initialize_buffers(struct asrc_farrow *src_obj)
+{
+	int32_t *buf_32;
+	int16_t *buf_16;
+	int ch;
+	size_t buffer_size;
+
+	/* Set buffer_length to filter_length * 2 to compensate for
+	 * missing element wise wrap around while loading but allowing
+	 * aligned loads.
+	 */
+	src_obj->buffer_length = src_obj->filter_length * 2;
+	src_obj->buffer_write_position = src_obj->filter_length;
+
+	if (src_obj->bit_depth == 32) {
+		buffer_size = src_obj->buffer_length * sizeof(int32_t);
+
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_32 = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, buffer_size);
+
+			if (!buf_32)
+				return -ENOMEM;
+
+			src_obj->ring_buffers32[ch] = buf_32;
+		}
+	} else {
+		buffer_size = src_obj->buffer_length * sizeof(int16_t);
+
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_16 = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, buffer_size);
+
+			if (!buf_16)
+				return -ENOMEM;
+
+			src_obj->ring_buffers16[ch] = buf_16;
+		}
+	}
+
+	return 0;
+}
+
+static void asrc_release_buffers(struct asrc_farrow *src_obj)
+{
+	int32_t *buf_32;
+	int16_t *buf_16;
+	int ch;
+
+	if (!src_obj)
+		return;
+
+	if (src_obj->bit_depth == 32)
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_32 = src_obj->ring_buffers32[ch];
+
+			if (buf_32) {
+				rfree(buf_32);
+				src_obj->ring_buffers32[ch] = NULL;
+			}
+		}
+	else
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_16 = src_obj->ring_buffers16[ch];
+
+			if (buf_16) {
+				rfree(buf_16);
+				src_obj->ring_buffers16[ch] = NULL;
+			}
+		}
+}
+
+static int asrc_dai_find(struct comp_dev *dev, struct comp_data *cd)
+{
+	struct comp_buffer *sourceb, *sinkb;
+	struct comp_dev *dev_dai;
+	int pid;
+
+	/* Get current pipeline ID and walk to find the DAI */
+	pid = dev_comp_pipe_id(dev);
+	cd->dai_dev = NULL;
+	if (cd->mode == ASRC_OM_PUSH) {
+		/* In push mode check if sink component is DAI */
+		do {
+			sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
```

Comment:
am I missing something or how is this advancing with loop iterations? Wouldn't it loop permanently if the first iteration doesn't find the DAI?